### PR TITLE
Add sensitivity_data column to instrument table

### LINKED ISSUE #2527: Instruments with limiting magnitude and associated exposure time
As part of the follow-up API, having some way of reporting a characteristic limiting magnitude for a given exposure time would be nice to then display when making a follow-up request, computing something like:

            nmag = -2.5*np.log10(np.sqrt(config_struct["exposuretime"]/exposureTime))
            limiting_mag = config_struct["magnitude"] + nmag

So adding this information as optional columns to the Instrument table (setting, for example, ZTF has 20.5 mag in 30 seconds).

### PR DESCRIPTION
**Pull request by jadalilleboe**
Closes #2527 

Adds a sensitivity_data column to the instrument table for filters and their associated limiting magnitude and exposure time.

### CODE REVIEW

**Comment by mcoughlin** [comment]
To be slightly more generic, I am wondering if perhaps we do something like a json packet that has the possibility of filter, magnitude, characteristic exposure time, as it will depend on the particular filter what those numbers are. Then these could be displayed on the front-end of the planning page when an instrument is chosen. What do you think @profjsb ?

File: skyportal/models/instrument.py [hunk 1]
```diff
@@ -88,6 +88,10 @@ class Instrument(Base):
         doc='List of filters on the instrument (if any).',
     )
 
+    mag = sa.Column(sa.Float, nullable=True, doc="Characteristic limiting magnitude")
```
Comment (profjsb) [inline skyportal/models/instrument.py:91]: maybe call this `limiting_mag`?

File: alembic/versions/f492e8ed9d84_add_magnitude_and_exposure_time_columns_.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+Revises: f9611048498a
+Create Date: 2022-01-27 10:06:33.987883
+
+"""
+from alembic import op
+import sqlalchemy as sa
+
+# revision identifiers, used by Alembic.
+revision = 'f492e8ed9d84'
+down_revision = '82f5a89fbe59'
```
Comment (profjsb) [inline alembic/versions/f492e8ed9d84_add_magnitude_and_exposure_time_columns_.py:13]: will need to update this down revision hash given that there are new migrations.

**Comment by profjsb** [comment]
Once this is out of Draft mode, please retag me for review.

**Comment by mcoughlin** [comment]
@jadalilleboe @profjsb Maybe we discuss at stand up on Thursday, but something more generic like a JSON that knows about filters and so on would probably be more future proof.

**Comment by jadalilleboe** [comment]
@mcoughlin So are you proposing to get rid of the filters, limiting_mag, and exposure_time columns and combine them all into one JSON column?

**Comment by mcoughlin** [comment]
@jadalilleboe yes (or similar). Can we discuss at stand-up?

**Comment by profjsb** [comment]
sensitivity_data ?

**Comment by jadalilleboe** [comment]
@mcoughlin @profjsb ok I think this is ready for review now

File: skyportal/handlers/api/instrument.py [hunk 3]
```diff
@@ -49,6 +49,15 @@ def post(self):
                 'Invalid/missing parameters: ' f'{exc.normalized_messages()}'
             )
 
+        sensitivity_data = data.pop("sensitivity_data", None)
+        if sensitivity_data:
+            filters = data.get("filters")
```
Comment (profjsb) [inline skyportal/handlers/api/instrument.py:54]: Will want to make sure filters is always a list
```suggestion
            filters = data.get("filters", [])
```

**Review by profjsb — CHANGES_REQUESTED** [review]
Very nice @jadalilleboe -- just a tiny change then good to go.

**Review by mcoughlin — APPROVED** [review]
Great @jadalilleboe. Can you confirm the down_revision is up to date? If so, happy to see this merged.

**Comment by jadalilleboe** [comment]
@mcoughlin yes can confirm it's up to date.

**Comment by jadalilleboe** [comment]
not anymore, will update later

File: alembic/versions/73bc8919f190_create_sensitivity_data_column.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+Create Date: 2022-02-16 18:31:00.739464
+
+"""
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+
+# revision identifiers, used by Alembic.
+revision = '73bc8919f190'
+down_revision = 'd6b24747693a'
```
Comment (profjsb) [inline alembic/versions/73bc8919f190_create_sensitivity_data_column.py:14]: check this...may have changed.

File: skyportal/tests/api/test_instrument.py [hunk 5]
```diff
@@ -7,6 +7,54 @@
...
+            ],
+            'telescope_id': telescope_id,
+            'field_data': pd.read_csv(fielddatafile)[:5].to_dict(orient='list'),
+            'field_region': Regions.read(regionsdatafile).serialize(format='ds9'),
+        },
+        token=super_admin_token,
+    )
+    assert status == 400
+    assert data['status'] == 'error'
+
```
Comment (profjsb) [inline skyportal/tests/api/test_instrument.py:56]: check that the error message is what you expect.

**Review by profjsb — CHANGES_REQUESTED** [review]
Looks good. A few more suggestions:
1) try testing when you add sensitivity_data with more than one entry.
2) can you add more sensitivity data to an existing instrument (via PATCH?) or does someone need to delete the entire instrument? I think it's a good idea to allow PATCHing of sensitivity_data. In this respect, it might be better to have `sensitivity_data` be a dict with the key as the (allowed) filter name. 

 ```python
'sensitivity_data': 
                { 'f110w': 
                   {
                    'limiting_magnitude': 20.5,
                    'magsys': 'ab',
                    'exposure_time': 30,
                   }
                }
```
3)  add a bit of info about the expected behaviors/data in the docs?

**Comment by mcoughlin** [comment]
@jadalilleboe Closing in favor of:
https://github.com/skyportal/skyportal/pull/2765
where we will apply these in. We can fix any feedback there.

**Pull request closed without merging**